### PR TITLE
drivers: can: socket: Use proper filter when setsockopt is called

### PR DESCRIPTION
**Pull request by jukkar**
We need to be prepared to receive either can_filter and zcan_filter
structs.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

### CODE REVIEW

**Comment by alexanderwachter** [comment]
I thought we only allow can_filter (Linux version)

**Comment by jukkar** [comment]
> I thought we only allow can_filter (Linux version)

At the moment, the socket-can sample app uses the zcan_filter when setting the filter values. In a way that is nicer to work with than can_filter as it is easier to set the individual fields in the filter struct.
But perhaps it makes sense to only allow can_filter to be used in this case, I just need to fix the socket-can sample application to only do Linux style filter setting.

**Comment by jukkar** [comment]
Updated the code so that
* User can only use "struct can_filter" when calling `setsockopt()`
* Generic `setsockopt()` code will check that the filter is correct size. I removed the checks from the driver as the socket code can do it in one place
* Modified the sample app to use can_filter when setting the filter. I left the zcan_filter still there in the app as it is much more convenient to use. The sample app will convert the filter when needed.

**Comment by jukkar** [comment]
The constness patch in this pr is coming actually from #14193 but as that PR is not yet merged and in order to avoid compile warning, I added it also to this PR.

File: drivers/can/socket_can_generic.h [hunk 1]
```diff
@@ -78,20 +78,21 @@ static inline int socket_can_setsockopt(struct device *dev, void *obj,
...
 	struct socket_can_context *socket_context = dev->driver_data;
+	struct zcan_filter filter;
 	int ret;
 
 	if (level != SOL_CAN_RAW && optname != CAN_RAW_FILTER) {
 		errno = EINVAL;
 		return -1;
 	}
 
-	if (optlen != sizeof(struct can_filter)) {
```
Comment (alexanderwachter) [inline drivers/can/socket_can_generic.h]: I would suggest an ASSERT optlen == sizeof(struct can_filter) here.

**Review by alexanderwachter — CHANGES_REQUESTED** [review]

File: drivers/can/socket_can_generic.h (hunk 1 above)
Comment (jukkar) [reply drivers/can/socket_can_generic.h]: We can add assert, no problem. It is just that the value is already checked also in sockets_can.c so we should not get into this function with wrong size address.

I noticed that we need to change the sending of the packet too, so "struct can_frame" should be used in sample when sending, then we need to convert it to "struct zcan_frame" as the driver is expecting that. I will add a commit for this in this PR.

**Comment by jukkar** [comment]
Updated the code:
* Added assert to generic socket-can driver code
* Moved can_filter -> zcan_filter conversion to sockets_can.c so that the driver does not have to do it
* Added can_frame -> zcan_frame conversion to sockets_can.c

Tested with nucleo_l432kc

**Review by alexanderwachter — APPROVED** [review]

**Review by rveerama1 — APPROVED** [review]

**Pull request merged**